File: web/src/lib/analytics.ts
// Umami analytics — privacy-friendly, self-hosted, gated on env vars.
//
// Init pattern is lifted from LaTabdhir (`LaTabdhir/front/src/lib/umami.ts`):
// when `VITE_UMAMI_URL` + `VITE_UMAMI_WEBSITE_ID` are both set, we inject
// the Umami `<script>` at boot. Otherwise the SDK never loads — every
// helper below becomes a silent no-op. Zero overhead in dev, zero
// privacy footprint when disabled.
//
// **Event policy.** The site is content-driven, not transactional, so we
// only track signals that answer real product questions:
//
// - page views (auto, on route change) — daily-return rhythm
// - {event,lesson,observance,person}_view — what people read
// - search — what people look for, with result count
// - language_change — trilingual reach signal
// - dispute_opened — engagement with the editorial dispute model
//
// We deliberately don't track theme toggles, filter clicks, pagination,
// or hover events — those are noise that bloat the dashboard without
// answering any decision.

const UMAMI_URL = import.meta.env.VITE_UMAMI_URL;
const UMAMI_WEBSITE_ID = import.meta.env.VITE_UMAMI_WEBSITE_ID;

interface UmamiPayload {
  hostname: string;
  language: string;
  referrer: string;
  screen: string;
  title: string;
  url: string;
  website: string;
}

// Umami's `track` is overloaded:
//   track()                                — record a page view at the current URL
//   track((props) => ({ ...props, url }))  — page view with a custom URL/title
//   track(name, data?)                     — custom named event
type UmamiTracker = {
  track: ((callback: (props: UmamiPayload) => object) => void) &
    ((eventName: string, data?: Record<string, unknown>) => void) &
    (() => void);
};

declare global {
  interface Window {
    umami?: UmamiTracker;
  }
}

/**
 * Inject the Umami script at boot. No-op when env vars aren't set.
 *
 * Call once from `main.tsx` before rendering. Subsequent calls are
 * idempotent — the function checks for an existing script tag.
 */
export function initAnalytics(): void {
  if (!UMAMI_URL || !UMAMI_WEBSITE_ID) return;
  if (document.getElementById("umami-script")) return;

  const script = document.createElement("script");
  script.id = "umami-script";
  script.async = true;
  script.defer = true;
  script.src = `${UMAMI_URL}/script.js`;
  script.setAttribute("data-website-id", UMAMI_WEBSITE_ID);
  // We track page views manually on route change (TanStack Router doesn't
  // trigger a real navigation), so disable Umami's auto-track.
  script.setAttribute("data-auto-track", "false");
  document.head.appendChild(script);
}

/**
 * Send a custom event. Silently no-ops when Umami isn't loaded —
 * either env vars aren't set, the script hasn't finished loading yet,
 * or a privacy extension is blocking it.
 */
function track(event: string, data?: Record<string, unknown>): void {
  window.umami?.track(event, data);
}

// ---------------------------------------------------------------------------
// Page views — call on every route change.
// ---------------------------------------------------------------------------

/**
 * Record a page view at ``path``. The Umami script may still be loading
 * when the first navigation happens; we retry briefly so the initial
 * landing isn't dropped.
 */
export function trackPageView(path: string, title?: string): void {
  if (!UMAMI_URL || !UMAMI_WEBSITE_ID) return;

  const send = () => {
    window.umami?.track((props) => ({
      ...props,
      url: path,
      title: title ?? document.title,
    }));
  };

  if (window.umami) {
    send();
    return;
  }

  // Poll for up to 5 seconds — Umami loads async and the first route
  // resolves before its `<script>` finishes parsing. Drop after that:
  // not worth keeping the interval alive longer for a page view.
  const interval = window.setInterval(() => {
    if (window.umami) {
      window.clearInterval(interval);
      send();
    }
  }, 200);
  window.setTimeout(() => window.clearInterval(interval), 5000);
}

// ---------------------------------------------------------------------------
// Domain-specific helpers. Exhaustively named so an IDE autocompletes
// the full menu of trackable signals.
// ---------------------------------------------------------------------------

/** User opened an event detail page. */
export function trackEventView(slug: string): void {
  track("event_view", { slug });
}

/** User opened a lesson detail page. */
export function trackLessonView(slug: string): void {
  track("lesson_view", { slug });
}

/** User opened an observance detail page. */
export function trackObservanceView(slug: string): void {
  track("observance_view", { slug });
}

/** User opened a person profile page. */
export function trackPersonView(slug: string): void {
  track("person_view", { slug });
}

/** Search query with the number of results returned (0 = no match). */
export function trackSearch(term: string, resultCount: number): void {
  // Trim and clip the term so we don't blow up Umami with novelty payloads
  // (a 4kb search box would otherwise dominate the events table).
  const trimmed = term.trim().slice(0, 80);
  if (!trimmed) return;
  track("search", { term: trimmed, results: resultCount });
}

/** User switched UI language. Signals real trilingual reach. */
export function trackLanguageChange(language: "en" | "fr" | "ar"): void {
  track("language_change", { language });
}

/** User opened the disputed-views drawer. Engagement with the editorial model. */
export function trackDisputeOpened(slug: string): void {
  track("dispute_opened", { slug });
}

File: mobile/lib/core/analytics/analytics.dart
/// Umami analytics for mobile — privacy-friendly, self-hosted, gated
/// on compile-time env vars.
///
/// Init pattern mirrors ``web/src/lib/analytics.ts``: when both
/// ``UMAMI_URL`` and ``UMAMI_WEBSITE_ID`` are passed via
/// ``--dart-define`` at build time, every helper here POSTs to
/// ``{UMAMI_URL}/api/send``. Without them, every helper is a silent
/// no-op — zero overhead in dev, zero privacy footprint when disabled.
///
/// Event policy (mirrors web):
///   - page_view (auto, on route change) — daily-return rhythm
///   - {event,lesson,observance,person}_view — what people read
///   - language_change — trilingual reach signal
///   - dispute_opened — engagement with editorial dispute model
///
/// Mobile-specific:
///   - bookmark_added — content-affinity signal
///   - auth_sign_up — funnel into accounts
///
/// Every payload includes ``data: {platform: "ios"|"android"}`` so the
/// shared Umami dashboard can split mobile vs. web traffic.
library;

import 'dart:async';
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';

class Analytics {
  Analytics._();

  static final Analytics instance = Analytics._();

  static const String _url = String.fromEnvironment('UMAMI_URL');
  static const String _websiteId = String.fromEnvironment('UMAMI_WEBSITE_ID');

  bool get _enabled => _url.isNotEmpty && _websiteId.isNotEmpty;

  /// Stable platform tag — splits mobile vs. web in the Umami dashboard.
  String get _platform {
    if (kIsWeb) return 'web';
    if (Platform.isIOS) return 'ios';
    if (Platform.isAndroid) return 'android';
    return 'other';
  }

  /// User-Agent used for every event. Umami infers sessions from
  /// IP + UA, so a consistent UA per install keeps the "unique
  /// visitor" count meaningful.
  String get _userAgent => 'Thaqafa-Mobile/$_platform';

  late final Dio _dio = Dio(BaseOptions(
    connectTimeout: const Duration(seconds: 3),
    receiveTimeout: const Duration(seconds: 3),
    headers: {
      'Content-Type': 'application/json',
      'User-Agent': _userAgent,
    },
  ));

  /// Send a payload to Umami's /api/send endpoint.
  ///
  /// **Failure-mode contract — analytics is *never* allowed to crash,
  /// block, or slow the host app.**
  ///
  /// Layers of defence, in order:
  ///   1. ``unawaited(...)`` — the caller does not await. The UI thread
  ///      doesn't wait for the network round-trip; the call returns the
  ///      moment the request is queued on Dio's HTTP client.
  ///   2. ``connectTimeout: 3s`` + ``receiveTimeout: 3s`` — bounded
  ///      wait. A hung connection is abandoned in 3 seconds; the user
  ///      never sees a stalled background task.
  ///   3. ``.catchError`` — silent. Every ``DioException`` (no internet,
  ///      DNS fail, TLS error, 5xx from Umami, timeout, …) is caught
  ///      and resolved into a no-op synthetic response. Nothing
  ///      propagates.
  ///   4. Outer ``try``/``catch`` — protects against synchronous
  ///      throws during payload construction or Dio's call set-up.
  ///      Belt and suspenders.
  ///   5. ``_enabled`` short-circuit — when ``UMAMI_URL`` /
  ///      ``UMAMI_WEBSITE_ID`` are unset, we don't even build the
  ///      payload, so zero overhead and zero network attempts.
  ///
  /// Net effect: the user can open the app on airplane mode, in the
  /// subway, behind a corporate proxy that blocks ``analytics.majlisna.app``,
  /// or in a country where the Umami host is unreachable — none of it
  /// is visible. The app keeps running; the event is silently lost.
  void _send({
    required String path,
    String? title,
    String? eventName,
    Map<String, dynamic>? eventData,
    String? language,
  }) {
    if (!_enabled) return;
    try {
      final payload = <String, dynamic>{
        'website': _websiteId,
        'hostname': 'thaqafa.app',
        'url': path,
        'language': language ?? '',
        'screen': '',
        'referrer': '',
        'title': ?title,
        'name': ?eventName,
        'data': {
          ...?eventData,
          'platform': _platform,
        },
      };
      unawaited(
        _dio
            .post<void>('$_url/api/send', data: {
          'type': 'event',
          'payload': payload,
        })
            .catchError((_) {
          // Silent: analytics never propagates failure to the host app.
          return Response<void>(requestOptions: RequestOptions());
        }),
      );
    } catch (_) {
      // Belt-and-suspenders: a synchronous throw during payload set-up
      // (shouldn't happen, but Dart's null-aware syntax + map literals
      // are evolving fast). Same contract — analytics never crashes.
    }
  }

  // ---------------------------------------------------------------------
  // Page views — call on every route change.
  // ---------------------------------------------------------------------

  /// Record a page view at ``path``. Called from the router observer
  /// after every navigation.
  void trackPageView(String path, {String? title, String? language}) =>
      _send(path: path, title: title, language: language);

  // ---------------------------------------------------------------------
  // Domain-specific helpers. Exhaustively named so an IDE autocompletes
  // the full menu of trackable signals.
  // ---------------------------------------------------------------------

  /// User opened an event detail page.
  void trackEventView(String slug, {String? language}) => _send(
        path: '/event/$slug',
        eventName: 'event_view',
        eventData: {'slug': slug},
        language: language,
      );

  /// User opened a lesson detail page.
  void trackLessonView(String slug, {String? language}) => _send(
        path: '/lesson/$slug',
        eventName: 'lesson_view',
        eventData: {'slug': slug},
        language: language,
      );

  /// User opened an observance detail page.
  void trackObservanceView(String slug, {String? language}) => _send(
        path: '/observance/$slug',
        eventName: 'observance_view',
        eventData: {'slug': slug},
        language: language,
      );

  /// User opened a person profile page.
  void trackPersonView(String slug, {String? language}) => _send(
        path: '/person/$slug',
        eventName: 'person_view',
        eventData: {'slug': slug},
        language: language,
      );

  /// User switched UI language. Real trilingual-reach signal.
  void trackLanguageChange(String language) => _send(
        path: '/settings',
        eventName: 'language_change',
        eventData: {'language': language},
        language: language,
      );

  /// User opened the disputed-views drawer.
  void trackDisputeOpened(String slug, {String? language}) => _send(
        path: '/event/$slug',
        eventName: 'dispute_opened',
        eventData: {'slug': slug},
        language: language,
      );

  /// User saved a bookmark. ``kind`` is one of event/lesson/observance/person.
  void trackBookmarkAdded(String kind, String slug, {String? language}) =>
      _send(
        path: '/bookmark',
        eventName: 'bookmark_added',
        eventData: {'kind': kind, 'slug': slug},
        language: language,
      );

  /// User completed sign-up. Funnel signal — paired with sign_in to see
  /// account creation vs. returning login.
  void trackSignUp({String? language}) => _send(
        path: '/sign-up',
        eventName: 'auth_sign_up',
        language: language,
      );

  /// User signed in to an existing account.
  void trackSignIn({String? language}) => _send(
        path: '/sign-in',
        eventName: 'auth_sign_in',
        language: language,
      );
}
